src/connector/src/sink/deltalake.rs

```diff
@@ -0,0 +1,544 @@
+    #[serde(rename = "s3.region")]
+    pub s3_region: Option<String>,
+    #[serde(rename = "s3.endpoint")]
+    pub s3_endpoint: Option<String>,
+}
+impl DeltaLakeCommon {
+    pub async fn create_deltalake_client(&self) -> Result<DeltaTable> {
+        let table = if let Some(local_path) = &self.local_path {
+            deltalake::open_table(local_path).await?
+        } else if let Some(s3_path) = &self.s3_path {
+            let mut storage_options = HashMap::new();
+            storage_options.insert(
+                AWS_ACCESS_KEY_ID.to_string(),
+                self.s3_access_key.clone().ok_or_else(|| {
+                    SinkError::Config(anyhow!("s3.access.key is required with aws s3"))
+                })?,
+            );
+            storage_options.insert(
+                AWS_SECRET_ACCESS_KEY.to_string(),
+                self.s3_secret_key.clone().ok_or_else(|| {
+                    SinkError::Config(anyhow!("s3.secret.key is required with aws s3"))
+                })?,
+            );
+            if let Some(s3_region) = &self.s3_region {
+                storage_options.insert(AWS_REGION.to_string(), s3_region.clone());
+            }
+            if let Some(s3_endpoint) = &self.s3_endpoint {
+                storage_options.insert(AWS_ENDPOINT_URL.to_string(), s3_endpoint.clone());
+            }
+            storage_options.insert(AWS_ALLOW_HTTP.to_string(), "true".to_string());
+            storage_options.insert(AWS_S3_ALLOW_UNSAFE_RENAME.to_string(), "true".to_string());
+            deltalake::open_table_with_storage_options(s3_path.clone(), storage_options).await?
+        } else {
+            return Err(SinkError::DeltaLake(
+                "`local.path` or `s3.path` set at least one, configure as needed.".to_string(),
+            ));
+        };
+        Ok(table)
+    }
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, WithOptions)]
+pub struct DeltaLakeConfig {
+    #[serde(flatten)]
+    pub common: DeltaLakeCommon,
+
+    pub r#type: String,
+}
+
+impl DeltaLakeConfig {
+    pub fn from_hashmap(properties: HashMap<String, String>) -> Result<Self> {
+        let config =
+            serde_json::from_value::<DeltaLakeConfig>(serde_json::to_value(properties).unwrap())
+                .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        if config.r#type != SINK_TYPE_APPEND_ONLY
+            && config.r#type != SINK_USER_FORCE_APPEND_ONLY_OPTION
+        {
+            return Err(SinkError::Config(anyhow!(
+                "only append-only delta lake sink is supported",
+            )));
+        }
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct DeltaLakeSink {
+    pub config: DeltaLakeConfig,
+    param: SinkParam,
+}
+
+impl DeltaLakeSink {
+    pub fn new(config: DeltaLakeConfig, param: SinkParam) -> Result<Self> {
+        Ok(Self { config, param })
+    }
+}
+
+fn check_field_type(rw_data_type: &DataType, dl_data_type: &DeltaLakeDataType) -> Result<bool> {
```

Comment:
Instead of return a bool, we can just include the unmatched case in the result and return `Result<()>`.